pkg/reporting/report.go

```diff
@@ -0,0 +1,251 @@
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+)
+
+const (
+	ReportSecretNameFormat = "kotsadm-%s-report"
+	ReportSecretKey        = "report"
+	ReportEventLimit       = 4000
+)
+
+var (
+	instanceReportMtx  = sync.Mutex{}
+	preflightReportMtx = sync.Mutex{}
+)
+
+type Report struct {
+	Events []ReportEvent `json:"events"`
+}
+
+type ReportEvent interface {
+	GetReportSecretName(appSlug string) string
+	GetReportSecretKey() string
+	GetReportEventLimit() int
+	GetReportMtx() *sync.Mutex
+	GetReportType() string
+}
+
+var _ ReportEvent = &InstanceReportEvent{}
+var _ ReportEvent = &PreflightReportEvent{}
+
+type InstanceReport struct {
+	Events []InstanceReportEvent `json:"events"`
+}
+
+type InstanceReportEvent struct {
+	ReportedAt                int64  `json:"reported_at"`
+	LicenseID                 string `json:"license_id"`
+	InstanceID                string `json:"instance_id"`
+	ClusterID                 string `json:"cluster_id"`
+	AppStatus                 string `json:"app_status"`
+	IsKurl                    bool   `json:"is_kurl"`
+	KurlNodeCountTotal        int    `json:"kurl_node_count_total"`
+	KurlNodeCountReady        int    `json:"kurl_node_count_ready"`
+	K8sVersion                string `json:"k8s_version"`
+	K8sDistribution           string `json:"k8s_distribution,omitempty"`
+	KotsVersion               string `json:"kots_version"`
+	KotsInstallID             string `json:"kots_install_id,omitempty"`
+	KurlInstallID             string `json:"kurl_install_id,omitempty"`
+	IsGitOpsEnabled           bool   `json:"is_gitops_enabled"`
+	GitOpsProvider            string `json:"gitops_provider"`
+	DownstreamChannelID       string `json:"downstream_channel_id,omitempty"`
+	DownstreamChannelSequence uint64 `json:"downstream_channel_sequence,omitempty"`
+	DownstreamChannelName     string `json:"downstream_channel_name,omitempty"`
+	DownstreamSequence        *int64 `json:"downstream_sequence,omitempty"`
+	DownstreamSource          string `json:"downstream_source,omitempty"`
+	InstallStatus             string `json:"install_status,omitempty"`
+	PreflightState            string `json:"preflight_state,omitempty"`
+	SkipPreflights            bool   `json:"skip_preflights"`
+	ReplHelmInstalls          int    `json:"repl_helm_installs"`
+	NativeHelmInstalls        int    `json:"native_helm_installs"`
+}
+
+func (i *InstanceReportEvent) GetReportType() string {
+	return "instance"
+}
+
+func (i *InstanceReportEvent) GetReportSecretName(appSlug string) string {
+	return fmt.Sprintf(ReportSecretNameFormat, fmt.Sprintf("%s-%s", appSlug, i.GetReportType()))
+}
+
+func (i *InstanceReportEvent) GetReportSecretKey() string {
+	return ReportSecretKey
+}
+
+func (i *InstanceReportEvent) GetReportEventLimit() int {
+	return ReportEventLimit
+}
+
+func (i *InstanceReportEvent) GetReportMtx() *sync.Mutex {
+	return &instanceReportMtx
+}
+
+type PreflightReport struct {
```

Comment:
i would also suggest moving preflight/instance specific types to its own files. e.g. `preflight_report.go` and `instance_report.go` along with the implementations of the interface functions.